### PR TITLE
Add CLI test - cli.import - test upload manifest

### PR DESCRIPTION
**Pull request by rplevka**
3 general changes:
- 1 added test (`test_import_orgs_manifests`)
    - the test utilizes the `robottelo.cli.manifests.clone()` method to generate fake manifests, renames them to match organization names (needed for `import` subcommand to be able to assign them to the appropriate orgs)
- 1 added method (`build_csv_file`)
    - generates a temporary file and feeds it by a header and records
    - more tests are coming up, where we need to craft custom csv files (we can possibly move the method to `common.helpers` or `cli.import_`)
- 1 modified method (`csv_to_dataset`)
    - the method now accepts the path to the csv instead of csv string itself. It now takes care of reading the file and parsing it.

```bash
$ nosetests -m test_import_orgs_manifests test_import.py
.
----------------------------------------------------------------------
Ran 1 test in 136.315s

OK
```

since I've modified some code used by `test_import_orgs_default`, here's the output for this test as well:
```bash
$ nosetests -m test_import_orgs_default test_import.py

.
----------------------------------------------------------------------
Ran 1 test in 14.701s

OK
```

### CODE REVIEW

File: tests/foreman/cli/test_import.py [hunk 1]
```diff
@@ -29,6 +36,16 @@ def csv_to_dataset(csv):
     ]
 
 
+def build_csv_file(header, rows=[]):
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Add a docstring. And for bonus karma, unit test this.

File: tests/foreman/cli/test_import.py [hunk 2]
```diff
@@ -70,6 +88,43 @@ def test_import_orgs_default(self):
...
+        """
+        files = prepare_import_data()[1]
+        csv_records = csv_to_dataset(files['users'])
+        # create number of manifests corresponding to the number of orgs
+        manifest_list = []
+        for org in set([rec['organization'] for rec in csv_records]):
+            for char in [' ', '.', '#']:
+                org = org.replace(char, '_')
+            man_file = manifests.clone()
+            ssh.upload_file(man_file, '/tmp/{}.zip'.format(org))
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Can `mktemp -d` be run on the remote server before uploading files? Doing so should help prevent collisions.

File: tests/foreman/cli/test_import.py [hunk 3]
```diff
@@ -70,6 +88,43 @@ def test_import_orgs_default(self):
...
+            # ssh.command(u'ls /tmp/{}.zip'.format(org))
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Is there any reason to keep this commented-out command?

File: tests/foreman/cli/test_import.py [hunk 4]
```diff
@@ -70,6 +88,43 @@ def test_import_orgs_default(self):
...
+            manifest_list.append('/tmp/{}.zip'.format(org))
+            os.remove(man_file)
+        ssh_import = Import.organization(
+            {'csv-file': files['users'], 'upload-manifests-from': '/tmp'}
+        )
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Formatting.

```python
        ssh_import = Import.organization({
            'csv-file': files['users'],
            'upload-manifests-from': '/tmp',
        })
```

File: tests/foreman/cli/test_import.py [hunk 5]
```diff
@@ -18,8 +19,14 @@
 
 
 @ddt
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Randomly - does anyone know why the `@ddt` decorator is applied to this method?

File: tests/foreman/cli/test_import.py (hunk 1 above)
Comment (rplevka) [inline tests/foreman/cli/test_import.py]: sure, no pr...waait a minute!

File: tests/foreman/cli/test_import.py [hunk 6]
```diff
@@ -56,7 +74,7 @@ def test_import_orgs_default(self):
         orgs = set(org['name'] for org in Org.list().stdout)
         imporgs = set(
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: "imporgs"?

File: tests/foreman/cli/test_import.py [hunk 7]
```diff
@@ -56,7 +74,7 @@ def test_import_orgs_default(self):
...
             org['organization'] for
-            org in csv_to_dataset(ssh_cat.stdout)
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: According to this change, the `ssh_import` and `ssh_cat` variables are no longer used. Drop them from this method.

File: tests/foreman/cli/test_import.py [hunk 8]
```diff
@@ -70,6 +88,43 @@ def test_import_orgs_default(self):
         )
         self.assertEqual(False in [org in orgs for org in imporgs], False)
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Can this assertion be rewritten somehow? How about this?

```python
self.assertTrue(all((org in orgs for org in imp_orgs)))
```

File: tests/foreman/cli/test_import.py (hunk 1 above)
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: The `mock` library provides good mocking facilities. You can even mock methods such as `open`. For some insight on the topic, see [How to Mock Python's `open` function](http://www.ichimonji10.name/blog/6/).

File: tests/foreman/cli/test_import.py [hunk 9]
```diff
@@ -18,8 +19,14 @@
 
 
 @ddt
-def csv_to_dataset(csv):
+def csv_to_dataset(csv_file):
     """Converts the csv with header as a first entry to a python dict"""
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Please expand this method's docstring.

**Comment by Ichimonji10** [comment]
ACK pending comments.

File: tests/foreman/cli/test_import.py (hunk 5 above)
Comment (elyezer) [inline tests/foreman/cli/test_import.py]: I think it was miss moved here as it should be on the class below.

Comment (elyezer) [inline tests/foreman/cli/test_import.py]: We need to update it!

File: tests/foreman/cli/test_import.py (hunk 8 above)
Comment (elyezer) [inline tests/foreman/cli/test_import.py]: :+1:

**Comment by elyezer** [comment]
ACK, pending @Ichimonji10's comments.

File: tests/foreman/cli/test_import.py (hunk 8 above)
Comment (rplevka) [inline tests/foreman/cli/test_import.py]: :+1:  ..much more elegant, will change it

File: tests/foreman/cli/test_import.py [hunk 10]
```diff
@@ -29,6 +36,23 @@ def csv_to_dataset(csv):
     ]
 
 
+def build_csv_file(rows=[]):
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: Do not use a mutable object as an argument to a function. You'll find lots of good explanations if you search for [python mutable argument](https://duckduckgo.com/?q=python+mutable+argument&t=ffsb&ia=qa). For example: http://stackoverflow.com/questions/1132941/least-astonishment-in-python-the-mutable-default-argument

File: tests/foreman/cli/test_import.py [hunk 11]
```diff
@@ -17,9 +18,15 @@
 from robottelo.cli.org import Org
 
 
-@ddt
-def csv_to_dataset(csv):
-    """Converts the csv with header as a first entry to a python dict"""
+def csv_to_dataset(csv_file):
+    """Takes a path to a remote csv file with a header, processes it
+    and returns its data represented by a dictionary object"""
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: How about something like this?

```
"""Process a remote CSV file.

Read a remote CSV file, process it and return it.

:param csv_file: A string. The path to a CSV file that resides on a remote server.
:returns: A dictionary holding the contents of the CSV file.

"""
```

File: tests/foreman/cli/test_import.py [hunk 12]
```diff
@@ -48,15 +73,12 @@ def test_import_orgs_default(self):
...
         files = prepare_import_data()[1]
         ssh_import = Import.organization({'csv-file': files['users']})
-        ssh_cat = ssh.command('cat {0}'.format(files['users']))
-        # pop the array as ssh.command appends 1 empty item at the end
-        ssh_cat.stdout.pop()
 
         # now to check whether the orgs from csv appeared in sattelite
         orgs = set(org['name'] for org in Org.list().stdout)
-        imporgs = set(
+        imp_orgs = set(
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: :+1:

File: tests/foreman/cli/test_import.py [hunk 13]
```diff
@@ -64,11 +86,47 @@ def test_import_orgs_default(self):
...
+            ssh.upload_file(man_file, '{0}/{1}.zip'.format(man_dir, org))
+            manifest_list.append('{0}/{1}.zip'.format(man_dir, org))
+            os.remove(man_file)
+        ssh_import = Import.organization({
+            'csv-file': files['users'],
+            'upload-manifests-from': man_dir,
+        })
+        # cleanup the file on remote and perform the assertions
+        ssh.command('rm -rf {}'.format(man_dir))
+        self.assertNotEqual(
```
Comment (Ichimonji10) [inline tests/foreman/cli/test_import.py]: How about this?

```python
self.assertIn('Created 3 organizations', ''.join(ssh_import.stdout))
```

There's two issues here:

* The `string.join` function only exists in Python 2. However, both Python 2 and 3 let you use `''.join()`. It's preferable to use the future-proofed approach.
* It's more straightforward to use the `assertIn` method than checking `find()`'s return code.

**Comment by Ichimonji10** [comment]
ACK pending comments.

**Comment by rplevka** [comment]
Comments addressed.

since assert functions have been updated:
```bash
$ nosetests -m test_import_orgs_manifests test_import.py
.
----------------------------------------------------------------------
Ran 1 test in 90.067s

OK
```

**Comment by Ichimonji10** [comment]
ACK

**Pull request merged**